### PR TITLE
[BUG Fixed] fix dist_op building when using PyTorch 2.0

### LINKED ISSUE #42: MS-AMP does not support pytorch2.0
**What's the issue, what's expected?**:
MS-AMP can not install with pytorch 2.0

**How to reproduce it?**:
Install MS-AMP in docker nvcr.io/nvidia/pytorch:23.01-py3.
```
sudo docker run -it  -d --name=torchtest --privileged --net=host --ipc=host --gpus=all nvcr.io/nvidia/pytorch:23.01-py3  bash
sudo docker exec -it torchtest bash

git clone https://github.com/Azure/MS-AMP.git
cd MS-AMP
python -m pip install --upgrade pip
pip install -e .
make postinstall
```

**Log message or shapshot?**:
image

**Additional information**:

### PR DESCRIPTION
**Pull request by wkcn**
**Description**
In PyTorch 2.0, when building `msamp_dist_op`, it raises the following error:
```
    /root/ms-amp/msamp/operators/dist_op/dist.cpp:209:38: error: no matching function for call to ‘pybind11::module_::def(const char [7], <unresolved overloaded function type>, const char [7])’
                                  209 |     m.def("reduce", &reduce, "Reduce");
                                      |                                      ^   
```

The cause is that there is a function called `reduce` in torch namespace. The function name is the same as that of our custom function.

In this PR, I rename the two functions `reduce` and `all_reduce`. Fix #42 

**Major Revision**
- rename two functions, i.e. `reduce` and `all_reduce` in `dist_op/dist.cpp`.
- remove `using namespace torch/std;` in dist_op
- link system nccl by adding `libraries=['nccl']` in setup.py

### CODE REVIEW

File: msamp/operators/dist_op/dist.cpp [hunk 1]
```diff
@@ -206,8 +205,8 @@ PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
                 return id;
             }));
     py::class_<ncclCommPtr>(m, "ncclCommPtr").def(py::init<>());
-    m.def("reduce", &reduce, "Reduce");
-    m.def("all_reduce", &all_reduce, "All reduce");
+    m.def("reduce", &dist_reduce, "Reduce");
+    m.def("all_reduce", &dist_all_reduce, "All reduce");
```
Comment (abuccts) [inline msamp/operators/dist_op/dist.cpp:209]: maybe rename to custom_reduce and custom_all_reduce?

**Review by abuccts — APPROVED** [review]
try to uncomment the following lines and run test with torch2.0
https://github.com/Azure/MS-AMP/blob/ac659d547bf380288bc1eb7b5e4f74839b954ceb/.github/workflows/unit-tests.yaml#L23-L24

**Review by guoshzhao — APPROVED** [review]

File: msamp/operators/dist_op/dist.cpp (hunk 1 above)
Comment (wkcn) [reply msamp/operators/dist_op/dist.cpp]: Done. I have renamed them.

**Comment by wkcn** [comment]
> try to uncomment the following lines and run test with torch2.0
> 
> https://github.com/Azure/MS-AMP/blob/ac659d547bf380288bc1eb7b5e4f74839b954ceb/.github/workflows/unit-tests.yaml#L23-L24

Got it.

**Comment by wkcn** [comment]
> try to uncomment the following lines and run test with torch2.0
> 
> https://github.com/Azure/MS-AMP/blob/ac659d547bf380288bc1eb7b5e4f74839b954ceb/.github/workflows/unit-tests.yaml#L23-L24

It raises the error in test with torch-2.0:
```
E           RuntimeError: The NVIDIA driver on your system is too old (found version 11080). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org/ to install a PyTorch version that has been compiled with your version of the CUDA driver.
```

**Comment by abuccts** [comment]
> > try to uncomment the following lines and run test with torch2.0
> > https://github.com/Azure/MS-AMP/blob/ac659d547bf380288bc1eb7b5e4f74839b954ceb/.github/workflows/unit-tests.yaml#L23-L24
> 
> It raises the error in test with torch-2.0:
> 
> ```
> E           RuntimeError: The NVIDIA driver on your system is too old (found version 11080). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org/ to install a PyTorch version that has been compiled with your version of the CUDA driver.
> ```

ok, maybe leave it as is for now and enable after upgrading host driver

**Pull request merged**